packages/clerk-js/src/core/clerk.ts

```diff
@@ -405,45 +403,39 @@ export default class Clerk implements ClerkInterface {
     // We now want to set the last active organization id on that session (if it exists).
     // However, if the `organization` parameter is not given (i.e. `undefined`), we want
     // to keep the organization id that the session had.
-    if (session && organization !== undefined) {
-      if (organization === null) {
-        session.lastActiveOrganizationId = null;
-      } else if (typeof organization === 'string') {
-        session.lastActiveOrganizationId = organization;
-      } else {
-        session.lastActiveOrganizationId = organization.id;
-      }
-
-      SessionTokenCache.clear();
+    const switchOrganization = organization !== undefined;
+    if (newSession && switchOrganization) {
+      const organizationId = typeof organization === 'string' ? organization : organization?.id;
+      newSession.lastActiveOrganizationId = organizationId || null;
     }
 
-    this.#authService?.setAuthCookiesFromSession(session);
-
     // If this.session exists, then signout was triggered by the current tab
     // and should emit. Other tabs should not emit the same event again
-    if (this.session && session === null) {
+    const signoutSession = this.session && newSession === null;
+    if (signoutSession) {
       this.#broadcastSignOutEvent();
     }
 
-    //1. setLastActiveSession to passed usersession (add a param).
+    //1. setLastActiveSession to passed user session (add a param).
     //   Note that this will also update the session's active organization
     //   id.
     if (inActiveBrowserTab()) {
-      await this.#touchLastActiveSession(session);
+      await this.#touchLastActiveSession(newSession);
+      // reload session from updated client
+      newSession = this.#getSessionFromClient(newSession.id);
     }
 
+    this.#authService?.setAuthCookiesFromSession(newSession);
```

Comment:
This will fire even when we're not in the currently active tab. I'm curious whether this can result in an out-of-sync cookie caused by having multiple tabs open.

In the case of calling `setAuthCookiesFromSession` within `setActive`:
- a user is a member of 2 orgs, org1 and org2
- the user opens tab1 and switches to org1. Tab1 shows: org1, cookie shows: org1
- the user opens tab2 and switches to org2. Tab2 shows: org2, cookie shows: org2, tab1 shows: org1

However, after the 2nd switch, tab1 still shows org1 as the selected org, even if the cookie shows org2 because it was overwritten after the switch.


In the case of calling `setAuthCookiesFromSession` from the poller: 
I need to explore this further, but can a poller fired from a different tab cause the cookie to fall out of sync?

This comment intends to trigger a discussion in case we require more changes in the setActive/cookie syncing algorithm - not a code change for now.

Reply:
syncing tabs session with global cookie value will always cause the cookie to have an incorrect value in the inactive tabs (if there are on different orgs or users). This is a problem of mutating a global resource based on a local resource.

Reply:
will discuss it offline in slack #team-js channel